File: typescript-effect/src/run-options.ts
import { Data } from "effect"
import type { HttpMethod } from "effect/unstable/http/HttpMethod"

export type Stop = Data.TaggedEnum<{
  Duration: { readonly seconds: number }
  Amount: { readonly requests: number }
}>

export const Stop = Data.taggedEnum<Stop>()

export type Body = Data.TaggedEnum<{
  Empty: { readonly empty?: never }
  Text: { readonly value: string }
}>

export const Body = Data.taggedEnum<Body>()

export type RunOptions = {
  readonly url: URL
  readonly connections: number
  readonly stop: Stop
  readonly method: HttpMethod
  readonly headers: Record<string, string>
  readonly body: Body
  readonly timeoutSeconds: number
}

File: go/runoptions/runoptions.go
package runoptions

import "net/url"

type Stop interface {
	stop()
}

type Duration struct {
	Seconds int
}

func (Duration) stop() {}

type Amount struct {
	Requests int
}

func (Amount) stop() {}

type Method string

const (
	MethodGet     Method = "GET"
	MethodPost    Method = "POST"
	MethodPut     Method = "PUT"
	MethodDelete  Method = "DELETE"
	MethodPatch   Method = "PATCH"
	MethodHead    Method = "HEAD"
	MethodOptions Method = "OPTIONS"
	MethodTrace   Method = "TRACE"
)

type RunOptions struct {
	URL            *url.URL
	Connections    int
	Stop           Stop
	Method         Method
	Headers        map[string]string
	Body           string
	TimeoutSeconds int
}
